packages/salesforcedx-vscode-soql/src/queryDataView/queryDataFileService.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { JsonMap } from '@salesforce/ts-types';
+import * as fs from 'fs';
+import { QueryResult } from 'jsforce';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { QUERY_RESULTS_DIR_PATH } from '../constants';
+import {
+  CsvDataProvider,
+  DataProvider,
+  JsonDataProvider
+} from './dataProviders';
+
+export enum FileFormat {
+  JSON = 'json',
+  CSV = 'csv'
+}
+
+const sfdxCoreExtension = vscode.extensions.getExtension(
+  'salesforce.salesforcedx-vscode-core'
+);
+const sfdxCoreExports = sfdxCoreExtension
+  ? sfdxCoreExtension.exports
+  : undefined;
+const { getRootWorkspacePath } = sfdxCoreExports;
+
+export class QueryDataFileService {
+  private dataProvider: DataProvider;
+
+  constructor(
+    private queryData: QueryResult<JsonMap>,
+    private format: FileFormat,
+    private documentName: string
+  ) {
+    this.dataProvider = this.getDataProvider();
+  }
+
+  protected getDataProvider(): DataProvider {
+    switch (this.format) {
+      case FileFormat.CSV:
+        return new CsvDataProvider(this.documentName);
+      case FileFormat.JSON:
+        return new JsonDataProvider(this.documentName);
+      default:
+        throw new Error('No DataProvider Found');
+    }
+  }
+
+  public save(): string {
+    try {
+      const fileContent = this.dataProvider.getFileContent(
+        this.queryData.records
+      );
+      const savedFileName = this.dataProvider.getFileName();
+      const queryDataFilePath = path.join(
+        this.getResultsDirectoryPath(),
+        savedFileName
+      );
+
+      this.createResultsDirectoryIfDoesNotExist();
+      // Save query results to disk
+      fs.writeFileSync(queryDataFilePath, fileContent);
+      this.showSaveSuccessMessage(savedFileName);
+      this.showFileInExplorer(queryDataFilePath);
+
+      return queryDataFilePath;
+    } catch (error) {
+      // TODO: i18n, CCX
+      vscode.window.showErrorMessage(
+        `Your data could not be saved. Run the query and try again.`
+      );
+      throw error;
+    }
+  }
+
+  private getResultsDirectoryPath() {
+    return path.join(getRootWorkspacePath(), QUERY_RESULTS_DIR_PATH);
+  }
+
+  private createResultsDirectoryIfDoesNotExist() {
+    fs.mkdirSync(this.getResultsDirectoryPath(), {
+      recursive: true
+    });
```

Comment:
Using the recursive option will create the directory if it does not exist already. So there is no need to check before creating it. Also, node [does not recommend](https://nodejs.org/api/fs.html#fs_fs_exists_path_callback) you check for existence anyway.

> Using fs.exists() to check for the existence of a file before calling fs.open(), fs.readFile() or fs.writeFile() is not recommended.